### PR TITLE
FEC workflow based on openFE/Alchemiscale

### PR DESCRIPTION
**Pull request by jthorton**
## Description
This PR adds an FEC setup and execution pipeline using OpenFE models in order to use Alchemiscale to run our calculations.

To track the provenance of our calculations and ensure reproducibility we use a schema which encodes all possible run time and setup settings. 

We also expose a CLI to make it easy to create and submit networks to alchemiscale. 

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [x] add results pull down once test calculations are complete 
  - [ ] 

## Questions
- [x] How to handle molecules with warts in the analysis?
- [x] other common alchemiscale actions we should expose?

## Status
- [ ] Ready to go

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ repos:
   - id: check-executables-have-shebangs
```
Comment (hmacdope) [inline .pre-commit-config.yaml:12]: Ah this is what is causing your PR to have more red, I'll move this to another PR and clean up there so its cleaner here.

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+        "Na+",
+        description="The positive monoatomic ion which should be used to neutralize the system and to adjust the ionic concentration.",
+    )
+    negative_ion: str = Field(
+        "Cl-",
+        description="The negative monoatomic ion which should be used to neutralize the system and to adjust the ionic concentration.",
+    )
+    neutralize: bool = Field(
+        True,
+        description="If the net charge of the chemical system should be neutralized by the ions defined by `positive_ion` abd `negative_ion`.",
```
Comment (hmacdope) [inline asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py:54]:
```suggestion
        description="If the net charge of the chemical system should be neutralized by the ions defined by `positive_ion` and `negative_ion`.",
```

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        for result in self.results:
+            raw_results[result.name()].append(result)
+
+        # make sure we have a solvent and complex phase for each result
+        for name, transforms in raw_results.items():
+            missing_phase = {"complex", "solvent"} - {t.phase for t in transforms}
+            if missing_phase:
+                raise RuntimeError(
+                    f"The transformation {name} is missing simulated legs in the following phases {missing_phase}"
+                )
```
Comment (hmacdope) [inline asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py:127]: As your require a 2 length iterable in   ` for leg1, leg2 in raw_results.values():` below, could also check that it is a 2 length iterable. I know this is a bit pedantic.

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+                leg1 if leg1.phase == "complex" else leg2
+            )
+            solvent_leg: TransformationResult = (
+                leg1 if leg1.phase == "solvent" else leg2
+            )
+            result = RelativeMeasurement(
+                labelA=leg1.ligand_a,
+                labelB=leg1.ligand_b,
+                DDG=(complex_leg.estimate - solvent_leg.estimate),
+                uncertainty=np.sqrt(
```
Comment (hmacdope) [inline asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py:142]: Make a function or  a note that this is for error propagation, may be useful elsewhere or at least to note to someone following along how its done.

File: asapdiscovery-simulation/asapdiscovery/simulation/utils.py [hunk 5]
```diff
@@ -20,3 +29,148 @@ def create_protein_only_system(input_pdb_path: str, ff_params: ForceFieldParams)
         hydrogenMass=ff_params.hydrogenMass,
     )
     return system
+
+
+class AlchemiscaleHelper:
+    """
+    A convince class to handle alchemiscale submissions restarts and results gathering.
```
Comment (hmacdope) [inline asapdiscovery-simulation/asapdiscovery/simulation/utils.py:36]:
```suggestion
    A convenience class to handle alchemiscale submissions restarts and results gathering.
```

File: asapdiscovery-simulation/asapdiscovery/simulation/utils.py [hunk 6]
```diff
@@ -20,3 +29,148 @@ def create_protein_only_system(input_pdb_path: str, ff_params: ForceFieldParams)
...
+            A list of actioned tasks for this network.
+
+        """
+        network_key = planned_network.results.network_key
+
+        tasks = []
+        for tf_sk in self._client.get_network_transformations(network_key):
+            # for each task create x repeats which are configured by the factory
+            tasks.extend(
+                self._client.create_tasks(tf_sk, count=planned_network.n_repeats + 1)
```
Comment (hmacdope) [inline asapdiscovery-simulation/asapdiscovery/simulation/utils.py:101]: Why is the +1 needed here?

File: asapdiscovery-simulation/asapdiscovery/simulation/utils.py [hunk 7]
```diff
@@ -20,3 +29,148 @@ def create_protein_only_system(input_pdb_path: str, ff_params: ForceFieldParams)
...
+            raw_result = self._client.get_transformation_results(tf_sk)
+            if raw_result is None:
+                continue
+            # format into our custom result schema and save
+            estimate = raw_result.get_estimate()
+            uncertainty = raw_result.get_uncertainty()
+            # work out the name of the molecules and the phase of the calculation
+            individual_runs = list(raw_result.data.values())
+
+            if "protein" in individual_runs[0][0].inputs["stateA"].components:
```
Comment (hmacdope) [inline asapdiscovery-simulation/asapdiscovery/simulation/utils.py:149]: Could you add a note as to why this is required? I couldn't figure it out just by looking.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looks great @jthorton! Thought I would go through quickly, but mostly just nitpicks/queries.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (hmacdope) [reply .pre-commit-config.yaml]: Fixed as of #431

**Comment by hmacdope** [comment]
Also might be good to detail if only in the interim why there are issues with frozen and non-frozen openFE stuff IIRC from reading through before.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (jthorton) [reply .pre-commit-config.yaml]: Thanks!

File: asapdiscovery-simulation/asapdiscovery/simulation/cli/plan.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        factory = FreeEnergyCalculationFactory.from_file(factory_file)
+
+    else:
+        factory = FreeEnergyCalculationFactory()
+
+    click.echo(f"Loading Ligands from {ligands}")
+    # parse all required data/ assume sdf currently
+    supplier = Chem.SDMolSupplier(ligands, removeHs=False)
+    input_ligands = [openfe.SmallMoleculeComponent.from_rdkit(mol) for mol in supplier]
+    if center_ligand is not None:
```
[the remark this answers is not in the document]
Comment (jthorton) [reply asapdiscovery-simulation/asapdiscovery/simulation/cli/plan.py:66]: Good catch!

File: asapdiscovery-simulation/asapdiscovery/simulation/cli/plan.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    click.echo("Creating FEC network ...")
+    planned_network = factory.create_fec_dataset(
+        dataset_name=name,
+        receptor=receptor,
+        ligands=input_ligands,
+        central_ligand=center_ligand,
+    )
+    click.echo(f"Writing results to {name}")
+    # output the data to a folder named after the dataset
+    os.makedirs(name, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (jthorton) [reply asapdiscovery-simulation/asapdiscovery/simulation/cli/plan.py:84]: Not really used it much but seems good, would you mind checking I have used it correctly :D

File: asapdiscovery-simulation/asapdiscovery/simulation/cli/submit.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    submitted_network = client.create_network(
+        planned_network=planned_network, scope=network_scope
+    )
+    # write the network with its key to file before we try and add compute incase we hit an issue
+    click.echo("Network made saving network key to network file")
+    submitted_network.to_file("planned_network.json")
+    # now action the tasks
+    click.echo("Creating and actioning FEC tasks on Alchemiscale...")
+    task_ids = client.action_network(planned_network=submitted_network)
+    # check that all tasks where created
```
[the remark this answers is not in the document]
Comment (jthorton) [reply asapdiscovery-simulation/asapdiscovery/simulation/cli/submit.py:70]: Well spotted!

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/atom_mapping.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+
+class LomapAtomMapper(_BaseAtomMapper):
+    """
+    A settings class for the LomapAtomMapper in openFE
+    """
+
+    type: Literal["LomapAtomMapper"] = "LomapAtomMapper"
+
+    time: PositiveInt = Field(
```
[the remark this answers is not in the document]
Comment (jthorton) [reply asapdiscovery-simulation/asapdiscovery/simulation/schema/atom_mapping.py:35]: Yeah, it currently matches the name in openfe but I agree I prefer it to be clear as well and its only a small change to get it in the correct format.

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/atom_mapping.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    preserve_chirality: bool = Field(
+        True,
+        description="If mappings must strictly preserve the chirality of the molecules.",
+    )
+    use_positions: bool = Field(
+        True,
+        description="If 3D positions should be used during the generation of the mappings.",
+    )
+    coordinate_tolerance: PositiveFloat = Field(
+        0.25,
```
[the remark this answers is not in the document]
Comment (jthorton) [reply asapdiscovery-simulation/asapdiscovery/simulation/schema/atom_mapping.py:93]: Yeah these are taken from the current openfe defaults.

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py (hunk 3 above)
Comment (jthorton) [reply asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py:127]: Good catch added a check for this as well and a separate test.

File: asapdiscovery-simulation/asapdiscovery/simulation/utils.py (hunk 6 above)
Comment (jthorton) [reply asapdiscovery-simulation/asapdiscovery/simulation/utils.py:101]: This is due to the factory recording how many times each transform should be repeated, so the total number of tasks is 1 + `n_repeats` we can change this field to be total runs if that would be more clear.

Comment (hmacdope) [reply asapdiscovery-simulation/asapdiscovery/simulation/utils.py:101]: happy either way

**Comment by hmacdope** [comment]
@jthorton I cleaned up so that the old tests are working again and not interfering. You will still need to sort out the rest of the deps but hopefully cleaner.

**Comment by JenkeScheen** [comment]
@ijpulidos could you take a look at the current default factory settings and let us know if anything needs adjusting?
```
{
  "type": "FreeEnergyCalculationFactory",
  "solvent_settings": {
    "type": "SolventSettings",
    "smiles": "O",
    "positive_ion": "Na+",
    "negative_ion": "Cl-",
    "neutralize": true,
    "ion_concentration": {
      "magnitude": 0.15,
      "unit": "molar",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    }
  },
  "forcefield_settings": {
    "constraints": "hbonds",
    "rigid_water": true,
    "remove_com": false,
    "hydrogen_mass": 3.0,
    "forcefields": [
      "amber/ff14SB.xml",
      "amber/tip3p_standard.xml",
      "amber/tip3p_HFE_multivalent.xml",
      "amber/phosaa10.xml"
    ],
    "small_molecule_forcefield": "openff-2.0.0"
  },
  "thermo_settings": {
    "temperature": {
      "magnitude": 298.15,
      "unit": "kelvin",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    },
    "pressure": {
      "magnitude": 0.9869232667160129,
      "unit": "standard_atmosphere",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    },
    "ph": null,
    "redox_potential": null
  },
  "system_settings": {
    "nonbonded_method": "PME",
    "nonbonded_cutoff": {
      "magnitude": 1.0,
      "unit": "nanometer",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    }
  },
  "solvation_settings": {
    "solvent_model": "tip3p",
    "solvent_padding": {
      "magnitude": 1.2,
      "unit": "nanometer",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    }
  },
  "alchemical_settings": {
    "lambda_functions": "default",
    "lambda_windows": 11,
    "unsampled_endstates": false,
    "use_dispersion_correction": false,
    "softcore_LJ_v2": true,
    "softcore_electrostatics": true,
    "softcore_alpha": 0.85,
    "softcore_electrostatics_alpha": 0.3,
    "softcore_sigma_Q": 1.0,
    "interpolate_old_and_new_14s": false,
    "flatten_torsions": false
  },
  "alchemical_sampler_settings": {
    "online_analysis_interval": 250,
    "n_repeats": 3,
    "sampler_method": "repex",
    "online_analysis_target_error": {
      "magnitude": 0.0,
      "unit": "boltzmann_constant * kelvin",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    },
    "online_analysis_minimum_iterations": 500,
    "flatness_criteria": "logZ-flatness",
    "gamma0": 1.0,
    "n_replicas": 11
  },
  "engine_settings": {
    "compute_platform": null
  },
  "integrator_settings": {
    "timestep": {
      "magnitude": 4,
      "unit": "femtosecond",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    },
    "collision_rate": {
      "magnitude": 1.0,
      "unit": "1 / picosecond",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    },
    "n_steps": {
      "magnitude": 250,
      "unit": "timestep",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    },
    "reassign_velocities": false,
    "n_restart_attempts": 20,
    "constraint_tolerance": 1e-06,
    "barostat_frequency": {
      "magnitude": 25,
      "unit": "timestep",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    }
  },
  "simulation_settings": {
    "equilibration_length": {
      "magnitude": 1.0,
      "unit": "nanosecond",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    },
    "production_length": {
      "magnitude": 5.0,
      "unit": "nanosecond",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    },
    "forcefield_cache": "db.json",
    "minimization_steps": 5000,
    "output_filename": "simulation.nc",
    "output_structure": "hybrid_system.pdb",
    "output_indices": "not water",
    "checkpoint_interval": {
      "magnitude": 250,
      "unit": "timestep",
      ":is_custom:": true,
      "pint_unit_registry": "openff_units"
    },
    "checkpoint_storage": "checkpoint.nc"
  },
  "protocol": "RelativeHybridTopologyProtocol",
  "n_repeats": 2,
  "network_planner": {
    "type": "NetworkPlanner",
    "atom_mapping_engine": {
      "type": "LomapAtomMapper",
      "timeout": 20,
      "threed": true,
      "max3d": 1000.0,
      "element_change": true,
      "seed": "",
      "shift": true
    },
    "scorer": "default_lomap",
    "network_planning_method": "minimal_spanning"
  }
}
```

**Comment by JenkeScheen** [comment]
@jthorton we should switch to `openff-2.1.0` if we can.

File: devtools/conda-envs/asapdiscovery.yaml [hunk 13]
```diff
@@ -83,3 +89,5 @@ dependencies:
     # viz
     - pygifsicle
     - sigfig
+    - git+https://github.com/openforcefield/alchemiscale.git@main
+    - git+https://github.com/OpenFreeEnergy/cinnabar.git@main
```
Comment (ijpulidos) [inline devtools/conda-envs/asapdiscovery.yaml:93]: Why is it that we need the git cinnabar and not the package at conda-forge? Do we need to push for a new release for it? :)

**Review by ijpulidos — CHANGES_REQUESTED** [review]
This is great work!

Just a few comments and one suggested change that should fix the CI tests not passing.

**Comment by ijpulidos** [comment]
@jthorton @JenkeScheen The settings look fine to me, I'd argue there's a little more solvent padding than needed and also we commonly use 12 replicas/windows instead of 11 (but I guess this is what the openfe folks have been using, I'd just double check with them). Also, I wonder what the difference is with `n_repeats` on the `alchemical_sampler_settings`, compared to the `n_repeats` later in the json/dictionary (basically there are two `n_repeats`, what is the difference between these?).

**Comment by ijpulidos** [comment]
Ah one thing that I just noticed is that the schema seems very oriented to the `openfe` relative bubdubg free energy protocol. As far as I remember we would like to be able to use other protocols (such as the NEQ cycling one from perses, or others that would be generated in the future), so I think we probably need to change some things for that. I think probably in the `FreeEnergyCalculationFactory`.

This might be something to improve/include in future iterations, but just wanted to bring this up.

File: devtools/conda-envs/asapdiscovery.yaml (hunk 13 above)
Comment (hmacdope) [reply devtools/conda-envs/asapdiscovery.yaml:93]: Yeah @ijpulidos this uses bleeding edge cinnabar code that isn't in the most recent conda package, so yes we should push for a release IMO if its not too much trouble.

File: asapdiscovery-simulation/asapdiscovery/simulation/tests/conftest.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    # convert to openfe objects
+    ligands = [openfe.SmallMoleculeComponent.from_rdkit(mol) for mol in supp]
+    return ligands
+
+
+@pytest.fixture(scope="session")
+def tyk2_protein():
+    """Create an openfe protein component"""
+    receptor_file = fetch_test_file("tyk2_protein.pdb")
+    return openfe.ProteinComponent.from_pdb_file(receptor_file.as_posix())
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-simulation/asapdiscovery/simulation/tests/conftest.py:22]: Live on S3 in the test-files bucket :)

File: devtools/conda-envs/asapdiscovery.yaml (hunk 13 above)
Comment (ijpulidos) [reply devtools/conda-envs/asapdiscovery.yaml:93]: Same holds for alchemiscale, actually

Comment (jthorton) [reply devtools/conda-envs/asapdiscovery.yaml:93]: As far as I could tell there is not an alchemiscale package on conda I might have missed that though.

**Comment by JenkeScheen** [comment]
currently testing the workflow, I think `rich` may be missing from the env.yml. On submit:
```
  File "/Users/jenkescheen/projects/fecs_workflow/asapdiscovery/asapdiscovery-simulation/asapdiscovery/simulation/cli/submit.py", line 61, in submit
    submitted_network = client.create_network(
  File "/Users/jenkescheen/projects/fecs_workflow/asapdiscovery/asapdiscovery-simulation/asapdiscovery/simulation/utils.py", line 73, in create_network
    network_key = self._client.create_network(fec_network, scope)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/alchemiscale/interface/client.py", line 111, in create_network
    from rich.progress import Progress
ModuleNotFoundError: No module named 'rich'
```

**Comment by jthorton** [comment]
Thanks for looking over it @ijpulidos. 
>  Also, I wonder what the difference is with n_repeats on the alchemical_sampler_settings, compared to the n_repeats later in the json/dictionary (basically there are two n_repeats, what is the difference between these?).

Good spot I missed that setting on the `alchemical_sampler_settings` it looks like we have two places where we can specify repeats which are doing the same thing. The setting on `alchemical_sampler_settings` replicates the jobs once submitted and `n_repeats` on the main schema is asking for repeat tasks when actioning them on alchemiscale. So it looks like we might be able to drop the extra field I introduced and just use the `alchemical_sampler_settings`. 

> schema seems very oriented to the openfe relative binding free energy protocol. As far as I remember we would like to be able to use other protocols (such as the NEQ cycling one from perses, or others that would be generated in the future), so I think we probably need to change some things for that.

Good point I imagined that the settings for the NEQ cycle method would have a lot of overlap with the current protocol so I broke it up into the component parts like `forcefield_settings` & `solvation_settings` etc with the expectation that we could then pass these to the other protocol. Do we have the other protocol somewhere where we can test how it will work with the factory? I checked on perses but couldn't find it on main.

File: asapdiscovery-simulation/asapdiscovery/simulation/tests/test_fec_schema.py [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+    # change the default settings to make sure they propagated
+    # change the lomap timeout
+    factory.network_planner.atom_mapping_engine.timeout = 30
+    factory.simulation_settings.equilibration_length = 0.5 * OFFUnit.nanoseconds
+    # plan a network
+    planned_network = factory.create_fec_dataset(
+        dataset_name="TYK2-test-dataset", receptor=tyk2_protein, ligands=tyk2_ligands
+    )
+    # make sure the settings were used correctly
+    assert planned_network.network.atom_mapping_engine.time == 30
```
[the remark this answers is not in the document]
Comment (jthorton) [reply asapdiscovery-simulation/asapdiscovery/simulation/tests/test_fec_schema.py:199]: great spot!

File: devtools/conda-envs/asapdiscovery.yaml (hunk 13 above)
Comment (ijpulidos) [reply devtools/conda-envs/asapdiscovery.yaml:94]: https://anaconda.org/openforcefield/alchemiscale-client

**Comment by ijpulidos** [comment]
> Good point I imagined that the settings for the NEQ cycle method would have a lot of overlap with the current protocol so I broke it up into the component parts like `forcefield_settings` & `solvation_settings` etc with the expectation that we could then pass these to the other protocol. Do we have the other protocol somewhere where we can test how it will work with the factory? I checked on perses but couldn't find it on main.

Yeah, we are still trying to get things into main, for now it lives in this branch https://github.com/choderalab/perses/tree/protocol-neqcyc sorry for that

**Comment by JenkeScheen** [comment]
[more testing:]

we might need to add a check that `planned_network.results != None`:
```
$ asap-alchemy gather
Connecting to Alchemiscale...
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/bin/asap-alchemy", line 8, in <module>
    sys.exit(cli())
  File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/jenkescheen/projects/fecs_workflow/asapdiscovery/asapdiscovery-simulation/asapdiscovery/simulation/cli/results.py", line 43, in gather
    status = client.network_status(planned_network=planned_network)
  File "/Users/jenkescheen/projects/fecs_workflow/asapdiscovery/asapdiscovery-simulation/asapdiscovery/simulation/utils.py", line 120, in network_status
    network_key = planned_network.results.network_key
AttributeError: 'NoneType' object has no attribute 'network_key'
```

File: devtools/conda-envs/asapdiscovery.yaml (hunk 13 above)
Comment (dotsdl) [reply devtools/conda-envs/asapdiscovery.yaml:94]: We do distribute on PyPI!

https://pypi.org/project/alchemiscale/

Sorry, forgot to push the 0.1.4 release; now it's there too!

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+    )
+    phase: Literal["complex", "solvent"] = Field(
+        ..., description="The phase of the transformation."
+    )
+    estimate: FloatQuantity["kcal/mol"] = Field(  # noqa: F821
+        ..., description="The average estimate of this transformation in kcal/mol"
+    )
+    uncertainty: FloatQuantity["kcal/mol"] = Field(  # noqa: F821
+        ...,
+        description="The standard deviation of the estimates of this transform in kcal/mol",
```
Comment (dotsdl) [inline asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py:86]: It is not generally the case that the `estimate` is an average. It is the case for the `openfe` `RelativeHybridTopologyProtocol`, where more repeats means more resolved distribution of the samples, with `estimate` as mean and `uncertainty` as standard deviation.

But it isn't the case for the `perses` `NonEquilibriumCyclingProtocol`. More repeats means more accumulated nonequilibrium work values, which are aggregated together to give a single, hopefully *more converged*, estimate. Uncertainty is calculated via bootstrap sampling of the accumulated nonequilibrium work values to quantify the distribution of possible estimates.

The most general way to document `estimate` and `uncertainty` here are that these are the "best" estimates for both the free energy difference and its uncertainty for the `Transformation` given the data.

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+                system_b = openfe.ChemicalSystem(
+                    sys_b_dict, name=f"{mapping.componentB.name}_{leg}"
+                )
+
+                transformation = openfe.Transformation(
+                    stateA=system_a,
+                    stateB=system_b,
+                    mapping={"ligand": mapping},
+                    protocol=protocol,  # use protocol created above
+                    name=f"{self.dataset_name}{system_a.name}_{system_b.name}",
```
Comment (dotsdl) [inline asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py:336]: Consider a separator between `{self.dataset_name}` and `{system_a.name}`.

**Review by dotsdl — COMMENTED** [review]
Looking pretty awesome so far @jthorton! Amazing work here!

File: devtools/conda-envs/asapdiscovery.yaml [hunk 18]
```diff
@@ -68,6 +68,13 @@ dependencies:
   - pymol-open-source
   - gifsicle
 
+  # simulation
+  - openfe >=0.11.0
+  - gufe >=0.9.1
```
Comment (dotsdl) [inline devtools/conda-envs/asapdiscovery.yaml]: Those should not be `>=`; we want `=` here.

**Comment by dotsdl** [comment]
> [more testing:]
> 
> we might need to add a check that `planned_network.results != None`:
> 
> ```
> $ asap-alchemy gather
> Connecting to Alchemiscale...
> Traceback (most recent call last):
>   File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/bin/asap-alchemy", line 8, in <module>
>     sys.exit(cli())
>   File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
>     return self.main(*args, **kwargs)
>   File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 1078, in main
>     rv = self.invoke(ctx)
>   File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/opt/homebrew/Caskroom/miniconda/base/envs/asapdiscovery_fecs/lib/python3.10/site-packages/click/core.py", line 783, in invoke
>     return __callback(*args, **kwargs)
>   File "/Users/jenkescheen/projects/fecs_workflow/asapdiscovery/asapdiscovery-simulation/asapdiscovery/simulation/cli/results.py", line 43, in gather
>     status = client.network_status(planned_network=planned_network)
>   File "/Users/jenkescheen/projects/fecs_workflow/asapdiscovery/asapdiscovery-simulation/asapdiscovery/simulation/utils.py", line 120, in network_status
>     network_key = planned_network.results.network_key
> AttributeError: 'NoneType' object has no attribute 'network_key'
> ```

I think I may have fixed the reason this can happen in 6307eb1917e447a6e9dab4b2bc41301151814bb1.

**Comment by JenkeScheen** [comment]
@dotsdl before merging this we should make sure there is an option for ensuring that all ligands in a generated network are in a cycle. Currently with `lomap`  (and obviously also minimal spanning trees) there is no way to make this happen, and we'd want to introduce some redundancy in production. I'm happy to work together to do this, it'll be a matter of diving deep into the `networkx` part of things.

**Comment by dotsdl** [comment]
Reviewed with @hmacdope; we *don't* believe there is any dependence on ligand names in the components added here. So, dealing with "warts" from OpenEye shouldn't have any bearing on behavior here.

**Review by dotsdl — APPROVED** [review]
Looks great @jthorton! Merging!

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py (hunk 16 above)
Comment (hmacdope) [reply asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py:86]: Followup issue #464

File: asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py (hunk 17 above)
Comment (hmacdope) [reply asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py:336]: Followup issue #465

**Review by hmacdope — APPROVED** [review]

**Comment by hmacdope** [comment]
Intermittent moonshot failure, merging.

**Pull request merged**